### DISCUSSION TITLE
spindle spinning at the wrong RPM

### DISCUSSION THREAD

**Discussion by tcurdt** [grblHAL/core#410] [Q&A]
I there can be many things that can be at fault here - and maybe it's not even GRBLHAL. But I am running out of ideas.
There original question I ask [over on discord](https://discord.com/channels/648972213734604807/1176075912362606682).

I have GRBLHAL 1.1f set to

```
$30 = 24000.000    (Maximum spindle speed, RPM)
$31 = 300.000    (Minimum spindle speed, RPM)
```

![vfd](https://github.com/grblHAL/core/assets/13697/c1e4a3f1-fd11-4950-a266-243bee26bd9f)
![spindle](https://github.com/grblHAL/core/assets/13697/0dbbe80c-33b6-4e2a-a955-9d48a6401dd8)

and with the Huanyang (HY 02D 223 B) set to:

```
PD003=400.00
PD004=400.00
PD005=400.00
PD144=3000 
```

it is reaching 400Hz (I measured about 24k RPM) with `M3 S7200` and not `M3 S24000`.

I am running out of ideas.

**Comment by tcurdt** [comment]
Do I maybe just need to adjust the RPM multipliers and dividers?

**Reply by terjeio** [reply]
Most Modbus controlled spindles do not use the $30 and $31 settings, the RPM range is read from the VFD.
Use the new [$SPINDLESH command](https://github.com/grblHAL/core/wiki/Report-extensions#spindles) to get at the [returned values](https://github.com/grblHAL/Plugins_spindle/blob/00e53219b24018ceef8c9f31dd19579ac0e5ec0b/vfd/huanyang.c#L228-L238).

So unless something is wrong with the Huanyang driver code you'll have to adjust the VFD settings.

**Reply by tcurdt** [reply]
Maybe the VFD needed another restart?
I honestly don't know why - but it's working now. And changed nothing 🤷‍♂️

Sorry for the noise.

**Reply by tcurdt** [reply]
Hm. I was cheering to early.

I had it complete working. `M3 S4000` and the VFD would display `RPM 4000`.

Today I turn everything on and suddenly it no longer is a 1:1 mapping between `M3 S<RPM>` and the spindle speed.
But his time it's even a different factor than before - and I did not touch the VFD config.

I somehow doubt it's a physical RS485 issue as I would expect checksum errors.
And for the given moment the gcode to speed commands are repeatable.

I am really puzzled.

Maybe I need to turn on one device before the other? otherwise it reads some wrong value or something?
I really have a hard time coming up with a scenario that could cause this behaviour.

Is there anything I could try or test?

**Reply by terjeio** [reply]
> Maybe I need to turn on one device before the other?

VFD first, or the controller cannot get the min/max RPM from it when starting up.
You should get alarm 14 if not first, if you clear the alarm before powering the VFD then min/max will not be read until the next soft reset.

**Reply by tcurdt** [reply]
Hm. OK.
That sounds like after a soft reset all should be fine no matter what.

Maybe this wasn't the case for older GRBLHAL versions?
(My GRBLHAL version is still from Jan because I could not figure out the settings for the web builder for a self build yet)

If there is no alarm 14 that means the low level communication should be fine, right?
So no need to wonder about cabling issues or something.
Just to rule that out.

**Reply by tcurdt** [reply]
Is there a way to see the min/max RPM that GRBLHAL reads via RS485?

Restarting the GRBLHAL board does not make a difference.
Even after a powering the board the values were repeatable.
Today it was stuck with these values:

2023-12-22:
```
S<rpm> rpm
 300  1000
1000  3000
2000  6600
3000 10000
4000 13000
5000 16000
6000 20000
```

There is the factor 3.33 again.

Does `GRBLHAL` just pass the information on? or is there the possibility for some kind of scaling?
I didn't find it in the code.

**Reply by terjeio** [reply]
> Is there a way to see the min/max RPM that GRBLHAL reads via RS485?

> Does GRBLHAL just pass the information on? or is there the possibility for some kind of scaling?

Follow the links in [this comment](https://github.com/grblHAL/core/discussions/410#discussioncomment-7904027).

**Reply by terjeio** [reply]
And you should update to the latest version, there has been changes to the code.
